File: Experimental/2021/LightingCppExample/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <frc/AddressableLED.h>
#include <frc/TimedRobot.h>
#include <frc/smartdashboard/SmartDashboard.h>

#include <array>

class Robot : public frc::TimedRobot {
  /// Maximum # of lights that we'll control (and the expected length of the
  /// strip, though something shorter *will* still function).
  static constexpr int kLength = 60;

  /// PWM port to which the LED strip is connected.
  ///
  /// Note that this *must* be a PWM header, not MXP or DIO.
  static constexpr int kLedPwmPortNum = 7;

  /// Software controller for addressable/digital LEDs.
  frc::AddressableLED m_led{kLedPwmPortNum};

  /// Buffer used to specify light values.  (Reusing a single buffer every time
  /// that we modify the lights, rather than constantly creating/releasing
  /// them.)
  std::array<frc::AddressableLED::LEDData, kLength> m_ledBuffer;

  /// Store what the last hue of the first pixel is.
  int firstPixelHue = 0;

 public:
  void RobotInit() override {
    // Default to a length of 60, start empty output
    // Length is expensive to set, so only set it once, then just update data
    m_led.SetLength(kLength);
    m_led.SetData(m_ledBuffer);
    m_led.Start();
  }

  void RobotPeriodic() override {
    // Fill the buffer with a rainbow
    UpdateLedBuffer();

    // Set the LEDs
    m_led.SetData(m_ledBuffer);
  }

 private:
  void UpdateLedBuffer() {
    // For every pixel
    for (int i = 0; i < kLength; i++) {
      // Calculate the hue - hue is easier for rainbows because the color
      // shape is a circle so only one value needs to precess
      const auto pixelHue = (firstPixelHue + (i * 180 / kLength)) % 180;
      // Set the value
      m_ledBuffer[i].SetHSV(pixelHue, 255, 128);
    }

    // Increase by to make the rainbow "move"
    firstPixelHue += 3;

    // Check bounds
    firstPixelHue %= 180;
  }
};

#ifndef RUNNING_FRC_TESTS
int main() {
  return frc::StartRobot<Robot>();
}
#endif

File: Experimental/2021/LightingExampleJava/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import edu.wpi.first.wpilibj.AddressableLED;
import edu.wpi.first.wpilibj.AddressableLEDBuffer;
import edu.wpi.first.wpilibj.TimedRobot;

public class Robot extends TimedRobot {
  private AddressableLED m_led;
  private AddressableLEDBuffer m_ledBuffer;
  // Store what the last hue of the first pixel is
  private int m_rainbowFirstPixelHue;

  // PWM port to which the LED strip is attached.
  // Must be a PWM header, not MXP or DIO
  static final int LED_PWM_PORT = 7;

  @Override
  public void robotInit() {
    System.out.println("Running robotInit()");

    m_led = new AddressableLED(LED_PWM_PORT);

    // Reuse buffer
    // Default to a length of 60
    // Length is expensive to set, so only set it once, then just update data
    m_ledBuffer = new AddressableLEDBuffer(60);
    m_led.setLength(m_ledBuffer.getLength());

    // Set the data
    m_led.setData(m_ledBuffer);
    m_led.start();
  }

  @Override
  public void disabledInit() {
    System.out.println("Disabled - setting LEDs to green");
    for (var i = 0; i < m_ledBuffer.getLength(); i++) {
      m_ledBuffer.setRGB(i, 0, 255, 0);
    }
    m_led.setData(m_ledBuffer);
  }

  @Override
  public void robotPeriodic() {
    // System.out.println("Periodic");
    // Fill the buffer with a rainbow
    rainbow();
    // Set the LEDs
    m_led.setData(m_ledBuffer);
  }

  private void rainbow() {
    // For every pixel
    for (var i = 0; i < m_ledBuffer.getLength(); i++) {
      // Calculate the hue - hue is easier for rainbows because the color
      // shape is a circle so only one value needs to precess
      final var hue = (m_rainbowFirstPixelHue + (i * 180 / m_ledBuffer.getLength())) % 180;
      // Set the value
      m_ledBuffer.setHSV(i, hue, 255, 128);
    }
    // Increase by to make the rainbow "move"
    m_rainbowFirstPixelHue += 3;
    // Check bounds
    m_rainbowFirstPixelHue %= 180;
  }
}
